### ISSUE TITLE
Is there a function to find the basename of all layers that are currently split?

### ISSUE THREAD

**Issue by bbimber** [satijalab/seurat#8960]
Hello,

The default behavior of Assay5 merged objects is to split layers. When one called JoinLayers(), we need to provide a list of layers to join. If we omit this, it defaults to counts/data/scale.data. That works fine unless your object has a non-standard layer name. 

LayerNames() returns the set of layer names, but unless I am missing something obvious it does not give a straightforward way to find the basenames of all layers. Stripping this list of numeric suffixes (i.e. counts.1 -> counts) isnt sufficient, since if you merge two seurat objects, the suffixes are project.name (i.e. counts.Proj1 and counts.Proj2). Splitting the layer names on '.' is also problematic since '.' isnt a unique delimiter (scale.data is a common layer name). 

Do you have any guidance on a strategy to take the result of LayerNames(), and find the names of all the layers that are currently split? Thanks for any help.

**Comment by samuel-marsh** [comment]
Hi,

Not member of team but hopefully can he helpful.  Can you explain your situation of what you are trying to do a little more?  Also just to clarify what is `LayerNames` function you are referring to as that is not Seurat function?

Best,
Sam

**Comment by bbimber** [comment]
@samuel-marsh: Sorry, the function is Layers(), not LayerNames(). I corrected the post above.

The core issue is that when seurat objected are merged in seurat5, the default is to create assays with split layers. This results an object with many layers, where they are named counts.1, counts.2, etc. Not all tools that operate on seurat objects understand how to interact with split layers. Thus, it frequently is important to join them. My lab runs many headless pipelines that run tools in different combinations, and therefore a robust way to completely undo the split is important. 

While seurat does supply a JoinLayers() function, one needs to know the names of the layers to join.  If one calls JoinLayers(layers = NULL), then the code will attempt to join the three standard layer names: counts, data, scale.data. Much of the time that's just fine. However, if your object contains a layer with a non-standard name, it will not be detected and not get joined. While a human can look at a list of layers and fairly easily infer the unsplit names for them, I am looking for code to automatically do this. 

The best I have is something we wrote:

```
.FindLayersToJoin <- function(seuratObj, assayName) {
	layerNames <- SeuratObject::Layers(seuratObj, assay = assayName)
	layerBasenames <- sapply(layerNames, function(x){
		x <- unlist(strsplit(x, split = '\\.'))
		if (length(x) == 1) {
			return(NA)
		}

		# Drop the final suffix, on the assumption the initial piece is the layer basename, and the final element is projectName:
		return(paste0(x[-length(x)], collapse = '.'))
	})

	# Any duplicated basenames inducate a split layer:
	layerBasenames <- layerBasenames[!is.na(layerBasenames)]
	layerBasenames <- unique(layerBasenames[duplicated(layerBasenames)])

	return(unlist(layerBasenames))
}
```

I'm simply asking whether there is some capability built into Seurat to more reliably perform JoinLayers() that I'm missing.

**Comment by samuel-marsh** [comment]
Gotcha.  Would the `search` parameter in `Layers` be what you need?

```
count_layers <- Layers(object = OBJ, search = "counts")
data_layers <- Layers(object = OBJ, search = "data")
```

This would return all count and data layers for merging with `JoinLayers` (leaving out scale data because you'll want that run over the whole object and then you can extract or drop suffixes as needed?

Or is it something else that you need?  

Best,
Sam

**Comment by bbimber** [comment]
Yes, I get that one can manually specify the terms. The point is to have code figure it out, rather than needing to know it upfront. Our usage involves a lot of headless pipelines, as opposed to a human making decisions.

**Comment by samuel-marsh** [comment]
Gotcha.

Just spitballing here but if the issue is knowing the names following an object merge then could you have pipeline change default assay settings prior to merge.
```
options(Seurat.object.assay.version = "v3")
```

That would merge things the old way.  You can then update the to Assay5 if needed and if you go to assay 5 and then split you would know the basenames because you specified the variable to split by?

**Comment by bbimber** [comment]
That's an interesting point worth considering. If I am reading the code right, this would apply to code automatically creating objects/assays, but would not change the behavior of merge() itself. Therefore it might prevent code from default to v5 assays, but if someone already has a seurat object with v5 assays, then merge() would still have the auto-split behavior.

**Comment by samuel-marsh** [comment]
oh ya that's true... I'll keep thinking but not sure I have any other ideas right now.  Hopefully Seurat Team can chime in with potential solution

**Comment by longmanz** [comment]
Hi,
Many thanks to @samuel-marsh for your help! 

To @bbimber, thank you for pointing this out. Unfortunately, we do not have such built-in function. Are you referring to cases where layers are not named with "counts.x" / "data.x"?  Normally, we use command like `count_layers <- Layers(object = OBJ, search = "counts")` to get all the "counts" / "data". Is there any way to force your pipeline to name layers using "counts.x" / "data.x"?

**Comment by bbimber** [comment]
@longmanz: thanks for the reply. Yes, I do understand how the search argument works. The ideal use case would involve:
```
seuratObj <- JoinLayers(seuratObj)
```
just automatically handling every split layer that is present and dealing with joining them without the user needing to think about anything. The existing JoinLayers() does this for the three standard layer names only. 

The function I posted above seems to basically work to infer which layers names represent split layers, but I could not find anything in a SeuratObject that is more explicit about layer relationships besides making inferences on the string layer names.
